codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/protocols/parsers/AwsQueryParserGenerator.kt

```diff
@@ -0,0 +1,87 @@
+
+/**
+ * The AWS query protocol's responses are identical to REST XML's, except that they are wrapped
+ * in a Response/Result tag pair:
+ *
+ * ```
+ * <SomeOperationResponse>
+ *     <SomeOperationResult>
+ *         <ActualData /> <!-- This part is the same as REST XML -->
+ *     </SomeOperationResult>
+ * </SomeOperationResponse>
+ * ```
+ *
+ * This class wraps [XmlBindingTraitParserGenerator] and uses it to render the vast majority
+ * of the response parsing, but it overrides [operationParser] to add the protocol differences.
+ */
+class AwsQueryParserGenerator(
+    protocolConfig: ProtocolConfig,
+    xmlErrors: RuntimeType,
+) : StructuredDataParserGenerator {
+    private val symbolProvider = protocolConfig.symbolProvider
+    private val smithyXml = CargoDependency.smithyXml(protocolConfig.runtimeConfig).asType()
+    private val xmlError = smithyXml.member("decode::XmlError")
+    private val xmlBindingGenerator = XmlBindingTraitParserGenerator(protocolConfig, xmlErrors)
+    private val codegenScope = arrayOf(
+        "Document" to smithyXml.member("decode::Document"),
+        "XmlError" to xmlError,
+    )
+
+    override fun payloadParser(member: MemberShape): RuntimeType = xmlBindingGenerator.payloadParser(member)
+    override fun errorParser(errorShape: StructureShape): RuntimeType? = xmlBindingGenerator.errorParser(errorShape)
+    override fun documentParser(operationShape: OperationShape): RuntimeType =
+        TODO("Document shapes are not supported by the AWS query protocol")
+
+    override fun operationParser(operationShape: OperationShape): RuntimeType? {
+        return xmlBindingGenerator.operationParserWithImpl(operationShape) { members ->
+            val operationName = symbolProvider.toSymbol(operationShape).name
+            val responseWrapperName = operationName + "Response"
+            val resultWrapperName = operationName + "Result"
+
+            rustTemplate(
+                """
+                use std::convert::TryFrom;
+                let mut doc = #{Document}::try_from(inp)?;
+
+                ##[allow(unused_mut)]
+                let mut decoder = doc.root_element()?;
+                let start_el = decoder.start_el();
+
+                if !(${XmlBindingTraitParserGenerator.XmlName(responseWrapperName).matchExpression("start_el")}) {
+                    return Err(#{XmlError}::custom(format!("invalid root, expected $responseWrapperName got {:?}", start_el)))
+                }
+                if let Some(mut result_tag) = decoder.next_tag() {
+                    let start_el = result_tag.start_el();
+                    if !(${XmlBindingTraitParserGenerator.XmlName(resultWrapperName).matchExpression("start_el")}) {
+                        return Err(#{XmlError}::custom(format!("invalid result, expected $resultWrapperName got {:?}", start_el)))
+                    }
+                """,
+                *codegenScope
+            )
+            xmlBindingGenerator.parseStructureInner(
```

Comment:
more as an fyi, you can use 
```kotlin
with(xmlBindingGenerator) {
 parseStructureInner()
}
```

if you wanted to keep the old function signature

Reply:
Thanks! TIL.